packages/bundle-source/cache.js

```diff
@@ -107,13 +107,38 @@ export const makeBundleCache = (wr, cwd, readPowers, opts) => {
   };
 
   /**
-   * @param {BundleMeta} meta
+   * @param {string} targetName
+   * @param {Logger} _log
+   * @returns {Promise<string>}
+   */
+  const loadMetaText = (targetName, _log = defaultLog) =>
+    wr
+      .readOnly()
+      .neighbor(toBundleMeta(targetName))
+      .maybeReadText()
+      .catch(
+        ioError =>
+          Fail`${targetName}: cannot read bundle metadata: ${q(ioError)}`,
+      );
+
+  /**
    * @param {string} targetName
    * @param {*} rootOpt
    * @param {Logger} [log]
+   * @param {BundleMeta} [meta]
    * @returns {Promise<BundleMeta>}
    */
-  const validate = async (meta, targetName, rootOpt, log = defaultLog) => {
+  const validate = async (
+    targetName,
+    rootOpt,
+    log = defaultLog,
+    meta = undefined,
+  ) => {
+    meta = await (meta ||
+      loadMetaText(targetName, log).then(metaText => JSON.parse(metaText)));
```

Comment:
Doesn't this violate our `await` rules in spirit, if not in fact? `loadMetaText` and `then` are both notionally capable of synchronous reentrancy/errors/etc.
```suggestion
    await null;
    if (!meta) {
      meta = await loadMetaText(targetName, log).then(metaText => JSON.parse(metaText));
    }
```

Reply:
We should also catch the JSON.parse error and annotate it with the source path.

Reply:
I don't think it violates the `await` spirit.  The subexpression is in the synchronous prelude, and the first `await` is still unconditional.

However, I'll rewrite it to be clearer anyway.  Thanks both!